dcli/lib/src/script/dart_project.dart

```diff
@@ -235,36 +228,43 @@ class DartProject {
   /// If [upgrade] is true then a pub upgrade is ran rather than
   /// pub get.
   ///
-  Future<void> warmup({bool background = false, bool upgrade = false}) async {
-    await _lock.withLock(
-      () async {
-        try {
-          if (background) {
-            // we run the clean in the background
-            // by running another copy of dcli.
-            print('DCli warmup started in the background.');
-            '${DCliPaths().dcliName} '
-                    '''-v=${join(io.Directory.systemTemp.path, 'dcli.warmup.log')}'''
-                    ' warmup $pathToProjectRoot'
-                .start(
-              detached: true,
-              runInShell: true,
-              extensionSearch: false,
-            );
-          } else {
-            // print(orange('Running pub get...'));
-            if (upgrade) {
-              await _pubupgrade();
+  void warmup({bool background = false, bool upgrade = false}) {
+    runtime.NamedLock.guard(
+      name: _lockName,
+      execution: runtime.ExecutionCall<void>(
+        callable: () {
+          try {
+            if (background) {
+              // we run the clean in the background
+              // by running another copy of dcli.
+              print('DCli warmup started in the background.');
+              '${DCliPaths().dcliName} '
+                      '''-v=${join(io.Directory.systemTemp.path, 'dcli.warmup.log')}'''
+                      ' warmup $pathToProjectRoot'
+                  .start(
+                detached: true,
+                runInShell: true,
+                extensionSearch: false,
+              );
             } else {
-              await _pubget();
+              // print(orange('Running pub get...'));
+              if (upgrade) {
+                _pubupgrade();
+              } else {
+                _pubget();
+              }
             }
+          } on PubGetException {
+            print(red("\ndcli warmup failed due to the 'pub get' call failing."));
           }
-        } on PubGetException {
-          print(red("\ndcli warmup failed due to the 'pub get' call failing."));
-        }
-      },
-      waiting: 'Waiting for warmup to complete...',
```

Comment:
Can definitely add something like this to the NamedLock.guard input parameters if we wanted.

Reply:
this comes back to my comments about retaining the existing NamedLocked API as part of our external API. This is nice little feature that makes it simple to provide user feedback.

Reply:
Hey @bsutton working on this as we speak! No problem here - can absolutely keep the API in tact!